### ISSUE TITLE
Detect useless assignment and return

### ISSUE THREAD

**Issue by tjwallace** [rubocop-hq/rubocop#7373]
Given this code, I'd like RuboCop to detect that the assignment is useless and autocorrect the code to remove the assignment.

```ruby
# bad
def foo(input)
  output = input.sort
  output
end

# good
def foo(input)
  input.sort
end
```

I'm not sure if this is a new cop or an addition to the existing `Link/UselessAssignment` cop.

**Comment by frostmark** [comment]
I will try to make it

**Comment by tjwallace** [comment]
> This issues been automatically closed due to lack of activity. Feel free to re-open it if you ever come back to it.

😞

**Comment by marcandre** [comment]
@tjwallace I don't particularly like stale bot either. You are free to reopen this.

Are you sure this is a pattern that happens enough to warrant a check?

Note that there can not be any code between `output = ...` and `output` (otherwise there could be side effects), so I think it's an easy node pattern, like `(... (lvasgn _name _expression) (lvar _name))`, call that `on_begin` (and alias `on_kwbegin`) and you have your cop...

**Comment by fatkodima** [comment]
Can confirm this is a popular pattern - I have seen those many times. Also I have found 3 offenses in `rails` codebase and 1 in `rubocop`s own.

**Comment by gurgeous** [comment]
Sorry, upgrading late here... This cop might be a bit aggressive. We have tons of data manipulation code like this:

```ruby
def normalize(input)
  string = input
  string = string.gsub(...)
  string = string.split.first
  string = string.upcase
  string
end
```

Stylistically there are many ways to approach this, but removing the final assignment is probably not helpful.  Thanks for the new rule, we would be happy to enable if it was slightly less aggressive.

**Comment by marcandre** [comment]
Maybe I'm taking your example too literally, but isn't it all much easier to understand without the `string`?

```ruby
def normalize(input)
  input
    .gsub(...)
    .split.first
    .upcase
end
```

**Comment by gurgeous** [comment]
Sometimes we use method chaining, but often we avoid it. I probably should've been more literal with my example. This is still made up, but more realistic. This pattern makes it easy to document, reorder, etc. Quite common with data processing IMO.

```ruby
def normalize(input)
  string = input

  # clean up whitespace
  string = string.gsub(...)

  # grab the first word
  string = string.split.first

  # normalize case to use as hash key
  string = string.upcase

  string
end
```

**Comment by marcandre** [comment]
Inline comments work well too, reordering is not an issue either... 😅

```ruby
def normalize(input)
  input
    .string.gsub(...) # clean up whitespace
    .split.first      # grab the first word
    .upcase           # normalize case to use as hash key
end
```
